### PR TITLE
New Component gist - manage gist in CLI

### PR DESCRIPTION
**Pull request by typebrook**
run `./gist/gist help` or just go to [`README`](https://github.com/typebrook/Bash-Snippets/tree/gist/gist)
for more information

**Pull Request Label:**
* [ ] Bug
* [ ] New feature
* [ ] Enhancement
* [x] New component
* [ ] Typo

**Pull Request Checklist:**
- [x] Have you followed the [guidelines for contributing](https://github.com/alexanderepstein/Bash-Snippets/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/alexanderepstein/Bash-Snippets/pulls) for the same fix or component?
- [x] Have you ran the tests locally with `bats tests`?

@alexanderepstein 
> Have you ran the tests locally with `bats tests`?

Yes, but since most of commands in this tool require Github API token (with gist scope), the tests I made didn't cover its core utilities. Maybe there is a better way to do the test? 

I promise this tool is really useful and user-friendly with prompts for toke validation and other stuffs, please give it a try!

### CODE REVIEW

**Comment by alexanderepstein** [comment]
I like this, it will most definitely make it to bash-snippets. I will need to take some time to read through it and review it before I can bring it in but overall it looks like it brings some nice functionality! Good work 👍

**Comment by typebrook** [comment]
@alexanderepstein 
Cool! Also, I made some improvement and `README` in latest commits.
Because I am new to bash, so if you find somewhere in code hard to read,  smells weird or violates some conventions, don't hesitate to ask me to do fix.

**Comment by typebrook** [comment]
@alexanderepstein 
Did some improvements.
Also, I add my api token from another Github account into test, so for now the tests should really meaningful for most of the sub-commands. But the weird thing is, all tests pass on my machine (Ubuntu 18.04), but test for `gist new` and `gist edit` never works on Travis build. Could you give it a try on your machine?

File: gist/README.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# gist - Manage your gist like a pro
```
Comment (alexanderepstein) [inline gist/README.md:1]: can you add a link to this readme under the main readme? If you look under the all tools section you should add an entry for gist and put the link to this readme.

File: install.sh [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env bash
 # Author: Alexander Epstein https://github.com/alexanderepstein
 currentVersion="1.23.0"
-declare -a tools=(bak2dvd bash-snippets cheat cloudup crypt cryptocurrency currency geo lyrics meme movies newton pwned qrify short siteciphers stocks taste todo transfer weather ytview)
+declare -a tools=(bak2dvd bash-snippets cheat cloudup crypt cryptocurrency currency geo lyrics meme movies newton pwned gist qrify short siteciphers stocks taste todo transfer weather ytview)
```
Comment (alexanderepstein) [inline install.sh:4]: Can we also add this to the uninstall script as well?

File: gist/gist [hunk 3]
```diff
@@ -0,0 +1,812 @@
...
+    done || exit 1
+
+    pin="${new_pinned[@]}"
+    _show_tags
+    sed -i'' -e "/^pin=/ d" "$CONFIG" && echo pin=\'"${new_pinned[@]}"\' >> "$CONFIG"
+  fi
+}
+
+_apply_config "$@" || exit 1
+getConfiguredClient || exit 1
```
Comment (alexanderepstein) [inline gist/gist:753]: Can we run checkInternet after this line? It seems all calls will require being able to connect to github.

File: gist/gist [hunk 4]
```diff
@@ -0,0 +1,812 @@
+#!/usr/bin/env bash
+#
+# Author: Hsieh Chin Fan (typebrook) <typebrook@gmail.com>
```
Comment (alexanderepstein) [inline gist/gist:3]: Can we add some of this to this [file](https://github.com/alexanderepstein/Bash-Snippets/blob/master/bash-snippets.1.ronn) and then use this [tool](https://github.com/rtomayko/ronn ) to generate this [file](https://github.com/alexanderepstein/Bash-Snippets/blob/master/bash-snippets.1)

File: gist/gist [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+               | sed -e '1 s/^/[/; $ s/$/]/' \
+               | _parse_response "${1#[[:alpha:]]}" )
+  [[ -n $new_record ]] && sed -i'' -E -e "/^$1 / s^.+^$new_record^" $INDEX \
+  && hint=false mark="$1 " _show_list \
+  || echo 'Fail to modify gist description'
+}
+
+# Print helper message
+usage() {
+  sed -E -n -e ' /^$/ q; 7,$ s/^# //p' "$0"
```
Comment (alexanderepstein) [inline gist/gist:672]: Can you clarify what is going on here? Most other tools just do 
```bash
  cat <<EOF
Usage goes here text goes here
EOF
```

**Comment by typebrook** [comment]
@alexanderepstein 
Thanks for your feedback, I'll resolve/answer them in these two days

File: gist/README.md (hunk 1 above)
Comment (typebrook) [reply gist/README.md:1]: fixed in d6d1343

File: install.sh (hunk 2 above)
Comment (typebrook) [reply install.sh]: fixed in 3e794b4

File: gist/gist (hunk 3 above)
Comment (typebrook) [reply gist/gist:804]: Just part of sub-commands(or sub-commands with proper arguments, like `gist tag <INDEX>`)  need internet, I added `checkInternet` to them.

check 21ce9c7

File: gist/gist (hunk 5 above)
Comment (typebrook) [reply gist/gist]: >Can you clarify what is going on here?

Print heading comments inside script. That is, the helper message (But exclude author information)

Compare to other tools, `gist` has too many parameters (sub-commands) and derivative　usages. Also, this script is also a gist on `gist.github.com`. That's why I prefer to put those necessary information at the top of script, so a user can get into it very quickly even it is not downloaded yet, or checking it in an editor. 

```
sed -E -n -e ' /^$/ q; 7,$ s/^# //p' "$0"
```
But specify a fixed line (7) in `sed` may not fit general use case. So I fixed it with pattern. check 4ec969f

File: gist/gist (hunk 4 above)
Comment (typebrook) [reply gist/gist:3]: Done in f9e14ca

**Comment by typebrook** [comment]
[`README`](https://github.com/typebrook/Bash-Snippets/tree/gist/gist) is up to date. Now it has index at the top. Also includes newly-added feature like `gist tag`, `gist lan` or pipe feature.

**Comment by typebrook** [comment]
@alexanderepstein 
Did some changes/improvements in the last few weeks:
1. Refactor/Bug fix for `gist tag` and the way to process description which are stored in local
2. Update [`README`](https://github.com/typebrook/Bash-Snippets/tree/gist/gist)
3. Mac compatibility
4. debug option for JSON parsing
5. For your latest review, you request to apply `checkInternet()` on every http request like other scripts.
    But since all request goes to `github.com` in this script, I just removed `checkInternet()` and integrated its logic into `http_method()`.

All your request are resolved, now you can review this PR again!

File: gist/gist (hunk 3 above)
Comment (typebrook) [reply gist/gist:835]: All http request go to `github.com`, so we can just check return code on every http request.
see 77868bb

**Comment by typebrook** [comment]
> Thanks for this, hopefully I will release a new version of bash-snippets soon and this will be part of that release!

Cool! Thanks for the review.

The [script I maintain on gist](https://gist.github.com/b0d2e7e67aa50298fdf8111ae7466b56)(which doesn't include Bash-Snippets stuff) is ahead of this PR with some refinement. I'll make another PR later to sync them.

**Pull request merged**